File: outbound-calls/typescript/advanced_example.ts
#!/usr/bin/env ts-node
/**
 * Advanced Voice Agent Testing Example - TypeScript
 * 
 * This example demonstrates how to:
 * 1. Create a test run and get assigned phone numbers
 * 2. Display the numbers to call
 * 3. Wait for completion and show detailed results
 * 4. Save results to file
 * 
 * Perfect for automated testing and CI/CD integration.
 */

import axios, { AxiosInstance } from 'axios';
import * as readline from 'readline';
import * as fs from 'fs';

interface TestRunResponse {
  testRunId: string;
  assignedNumbers: Array<{
    phoneNumber: string;
    testCaseTitle: string;
    testCaseRunId?: string;
  }>;
  expiresAt: string;
}

interface TestResults {
  summary: {
    id: string;
    status: string;
    stats: {
      total: number;
      completed: number;
      failed: number;
      pending: number;
      inProgress: number;
    };
  };
  results: Array<{
    id: string;
    testCaseId: string;
    status: string;
    durationSeconds?: number;
    recordingUrl?: string;
    transcriptionDataUrl?: string;
    analysis?: {
      overall_score?: number;
      key_points?: string[];
    };
  }>;
}

class HammingVoiceAgentAPI {
  private client: AxiosInstance;
  public baseUrl: string;

  constructor(apiKey: string, baseUrl: string = 'https://app.hamming.ai') {
    this.baseUrl = baseUrl;
    this.client = axios.create({
      baseURL: baseUrl,
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      }
    });
  }

  async createTestRun(
    agentId: string,
    testName?: string,
    tagIds?: string[],
    timeoutMinutes: number = 10
  ): Promise<TestRunResponse> {
    const response = await this.client.post('/api/rest/test-runs/test-outbound-agent', {
      agentId,
      name: testName || 'TypeScript API Test',
      timeoutMinutes,
      tagIds: tagIds || ['default']
    });
    
    return response.data;
  }

  async getTestRunStatus(testRunId: string): Promise<any> {
    const response = await this.client.get(`/api/rest/test-runs/${testRunId}/status`);
    return response.data;
  }

  async getTestResults(testRunId: string): Promise<TestResults> {
    const response = await this.client.get(`/api/rest/test-runs/${testRunId}/results`);
    return response.data;
  }
}

function printAssignedNumbers(assignedNumbers: TestRunResponse['assignedNumbers']): void {
  console.log('\n📞 ASSIGNED PHONE NUMBERS');
  console.log('='.repeat(60));
  console.log('Call these numbers to test your agent:');
  
  assignedNumbers.forEach((assignment, index) => {
    console.log(`  ${index + 1}. ${assignment.phoneNumber}`);
    console.log(`     Test Case: ${assignment.testCaseTitle}`);
    console.log(`     Test ID: ${assignment.testCaseRunId || 'N/A'}`);
    console.log();
  });
  
  console.log('📋 Instructions:');
  console.log('  1. Call each number from your test phone');
  console.log('  2. Follow the test case scenario');
  console.log('  3. Wait for the test run to complete');
  console.log('  4. Review results in the dashboard');
  console.log('='.repeat(60));
}

function printTestResults(results: TestResults, testRunId: string, dashboardUrl: string): void {
  console.log('\n' + '='.repeat(60));
  console.log('VOICE AGENT TEST RESULTS');
  console.log('='.repeat(60));
  console.log(`Test Run ID: ${testRunId}`);
  console.log(`View Results: ${dashboardUrl}`);
  
  const { stats } = results.summary;
  console.log(`Total Tests: ${stats.total}`);
  console.log(`Completed: ${stats.completed}`);
  console.log(`Failed: ${stats.failed}`);
  console.log(`Pending: ${stats.pending}`);
  
  if (stats.completed > 0) {
    const successRate = ((stats.completed - stats.failed) / stats.completed) * 100;
    console.log(`Success Rate: ${successRate.toFixed(1)}%`);
  }
  
  // Show individual test results
  if (results.results.length > 0) {
    console.log(`\nIndividual Test Results (${results.results.length} tests):`);
    results.results.forEach((result, index) => {
      console.log(`  ${index + 1}. Status: ${result.status}, Duration: ${result.durationSeconds || 0}s`);
      
      if (result.recordingUrl) {
        console.log(`     🎵 Recording: ${result.recordingUrl}`);
      }
      
      if (result.transcriptionDataUrl) {
        console.log(`     📝 Transcript: ${result.transcriptionDataUrl}`);
      }
      
      if (result.analysis) {
        console.log(`     Analysis Score: ${result.analysis.overall_score || 'N/A'}`);
        if (result.analysis.key_points) {
          console.log(`     Key Points: ${result.analysis.key_points.slice(0, 3).join(', ')}`);
        }
      }
    });
  }
  
  console.log('='.repeat(60));
}

async function waitForCompletion(api: HammingVoiceAgentAPI, testRunId: string, maxWaitSeconds: number = 600): Promise<TestResults | null> {
  console.log(`\n⏳ Waiting for test run ${testRunId} to complete...`);
  console.log('💡 Tip: You can skip waiting and check results later in the dashboard');
  
  const startTime = Date.now();
  
  while (true) {
    if (Date.now() - startTime > maxWaitSeconds * 1000) {
      console.log(`⏰ Timeout reached (${maxWaitSeconds}s). Check dashboard for results.`);
      return null;
    }
    
    const status = await api.getTestRunStatus(testRunId);
    
    if (['COMPLETED', 'FAILED'].includes(status.status)) {
      console.log(`✅ Test run ${status.status.toLowerCase()}`);
      break;
    }
    
    console.log(`🔄 Status: ${status.status} - waiting...`);
    await new Promise(resolve => setTimeout(resolve, 10000));
  }
  
  return await api.getTestResults(testRunId);
}

function askQuestion(question: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  return new Promise(resolve => {
    rl.question(question, answer => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

async function main(): Promise<number> {
  console.log('🚀 Starting Advanced Voice Agent Test');
  console.log('='.repeat(50));
  
  // Configuration - update these values
  const apiKey = process.env.HAMMING_API_KEY || 'your-api-key-here';
  const agentId = process.env.HAMMING_AGENT_ID || 'your-agent-id';
  const tagIds = ['default', 'api-test']; // Specify which test cases to run
  
  if (apiKey === 'your-api-key-here' || agentId === 'your-agent-id') {
    console.log('⚠️  Please set HAMMING_API_KEY and HAMMING_AGENT_ID environment variables');
    console.log('   Or update the values in the script');
    return 1;
  }
  
  try {
    const api = new HammingVoiceAgentAPI(apiKey);
    console.log('✅ API client initialized');
    
    // Create test run
    console.log('\n🔄 Creating test run...');
    const testRunResponse = await api.createTestRun(
      agentId,
      'Advanced TypeScript API Test',
      tagIds
    );
    
    const testRunId = testRunResponse.testRunId;
    const assignedNumbers = testRunResponse.assignedNumbers;
    
    console.log('✅ Test run created successfully!');
    console.log(`📊 Test Run ID: ${testRunId}`);
    
    // Display assigned numbers
    printAssignedNumbers(assignedNumbers);
    
    // Ask user if they want to wait for completion
    console.log('\n🤔 Options:');
    console.log('  1. Wait for test completion (recommended for small tests)');
    console.log('  2. Exit now and check results later in dashboard');
    
    const choice = await askQuestion('\nEnter choice (1 or 2): ');
    
    if (choice === '1') {
      // Monitor the test execution
      console.log('\n⏳ Monitoring test execution...');
      const results = await waitForCompletion(api, testRunId);
      if (results) {
        console.log('\n🎉 Test completed!');
        console.log(`📊 Final Status: ${results.summary?.status || 'Unknown'}`);
        
        // Get detailed results
        console.log('\n📊 Retrieving detailed results...');
        
        // Display results summary
        console.log('\n' + '='.repeat(50));
        console.log('📋 TEST RESULTS SUMMARY');
        console.log('='.repeat(50));
        
        const stats = results.summary?.stats || {};
        
        console.log(`Total Calls: ${stats.total || 0}`);
        console.log(`Successful Calls: ${(stats.completed || 0) - (stats.failed || 0)}`);
        console.log(`Failed Calls: ${stats.failed || 0}`);
        console.log(`Pending Calls: ${stats.pending || 0}`);
        
        // Show individual call results
        const callResults = results.results || [];
        if (callResults.length > 0) {
          console.log('\n📞 INDIVIDUAL CALL RESULTS:');
          callResults.forEach((call, index) => {
            console.log(`\n  Call ${index + 1}:`);
            console.log(`    Status: ${call.status || 'N/A'}`);
            console.log(`    Duration: ${(call.durationSeconds || 0).toFixed(1)}s`);
            
            if (call.recordingUrl) {
              console.log(`    🎵 Recording: ${call.recordingUrl}`);
            }
            
            if (call.transcriptionDataUrl) {
              console.log(`    📝 Transcript: ${call.transcriptionDataUrl}`);
            }
            
            if (call.analysis) {
              console.log(`    Analysis Score: ${call.analysis.overall_score || 'N/A'}`);
              if (call.analysis.key_points) {
                console.log(`    Key Points: ${call.analysis.key_points.slice(0, 3).join(', ')}`);
              }
            }
          });
        }
        
        // Save results to file
        const resultsFile = `outbound_test_results_${testRunId}.json`;
        fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2));
        
        console.log(`\n💾 Full results saved to: ${resultsFile}`);
        console.log('\n🎉 Test completed successfully!');
      }
    } else {
      // Just show dashboard link
      const dashboardUrl = `${api.baseUrl}/test-runs/${testRunId}`;
      console.log('\n✅ Test run created successfully!');
      console.log(`📊 Dashboard URL: ${dashboardUrl}`);
      console.log('📞 Call the numbers shown above to execute the tests');
      console.log('💡 Tip: Monitor progress in the dashboard');
    }
    
  } catch (error) {
    console.error(`\n❌ Error: ${(error as Error).message}`);
    return 1;
  }
  
  return 0;
}

// Run the example
if (require.main === module) {
  main().then(process.exit).catch(console.error);
}

File: outbound-calls/python/advanced_example.py
#!/usr/bin/env python3
"""
Advanced Outbound Call Testing Example

This example demonstrates how to:
1. Create a test run and get assigned phone numbers
2. Display the numbers to call
3. Wait for completion and show detailed results
4. Save results to file

Perfect for automated testing and CI/CD integration.
"""

import asyncio
import aiohttp
import json
import os
import time
from typing import Dict, Any, Optional, List

class HammingOutboundCallAPI:
    def __init__(self, api_key: str, base_url: str = "https://app.hamming.ai"):
        self.api_key = api_key
        self.base_url = base_url
        self.session: Optional[aiohttp.ClientSession] = None
    
    async def __aenter__(self):
        self.session = aiohttp.ClientSession(
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
        )
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()
    
    async def create_test_run(
        self,
        agent_id: str,
        test_name: Optional[str] = None,
        tag_ids: Optional[List[str]] = None,
        timeout_minutes: int = 10
    ) -> Dict[str, Any]:
        """Create a test run and get assigned phone numbers to call."""
        url = f"{self.base_url}/api/rest/test-runs/test-outbound-agent"
        payload = {
            "agentId": agent_id,
            "name": test_name or "API Test Run",
            "timeoutMinutes": timeout_minutes,
            "tagIds": tag_ids or ["default"]
        }
        
        async with self.session.post(url, json=payload) as response:
            if response.status != 200:
                error_text = await response.text()
                raise Exception(f"Failed to create test run: {response.status} - {error_text}")
            return await response.json()
    
    async def get_test_run_status(self, test_run_id: str) -> Dict[str, Any]:
        """Get current status of a test run."""
        url = f"{self.base_url}/api/rest/test-runs/{test_run_id}/status"
        async with self.session.get(url) as response:
            if response.status != 200:
                error_text = await response.text()
                raise Exception(f"Failed to get test run status: {response.status} - {error_text}")
            return await response.json()
    
    async def get_test_results(self, test_run_id: str) -> Dict[str, Any]:
        """Get detailed test results."""
        url = f"{self.base_url}/api/rest/test-runs/{test_run_id}/results"
        async with self.session.get(url) as response:
            if response.status != 200:
                error_text = await response.text()
                raise Exception(f"Failed to get test results: {response.status} - {error_text}")
            return await response.json()

def print_assigned_numbers(assigned_numbers: list):
    """Print formatted assigned phone numbers."""
    print(f"\n📞 ASSIGNED PHONE NUMBERS")
    print("="*60)
    print("Call these numbers to test your agent:")
    
    for i, assignment in enumerate(assigned_numbers, 1):
        phone_number = assignment.get('phoneNumber', 'N/A')
        test_case = assignment.get('testCaseTitle', 'Unknown Test')
        test_id = assignment.get('testCaseRunId', 'N/A')
        
        print(f"  {i}. {phone_number}")
        print(f"     Test Case: {test_case}")
        print(f"     Test ID: {test_id}")
        print()
    
    print("📋 Instructions:")
    print("  1. Call each number from your test phone")
    print("  2. Follow the test case scenario")
    print("  3. Wait for the test run to complete")
    print("  4. Review results in the dashboard")
    print("="*60)

def print_test_results(results: Dict[str, Any], test_run_id: str, dashboard_url: str):
    """Print formatted test results."""
    print("\n" + "="*60)
    print("VOICE AGENT TEST RESULTS")
    print("="*60)
    print(f"Test Run ID: {test_run_id}")
    print(f"View Results: {dashboard_url}")
    
    summary = results.get('summary', {})
    total_tests = summary.get('total', 0)
    completed_tests = summary.get('completed', 0)
    failed_tests = summary.get('failed', 0)
    pending_tests = summary.get('pending', 0)
    
    print(f"Total Tests: {total_tests}")
    print(f"Completed: {completed_tests}")
    print(f"Failed: {failed_tests}")
    print(f"Pending: {pending_tests}")
    
    if completed_tests > 0:
        success_rate = ((completed_tests - failed_tests) / completed_tests) * 100
        print(f"Success Rate: {success_rate:.1f}%")
    
    # Show individual test results
    test_results = results.get('results', [])
    if test_results:
        print(f"\nIndividual Test Results ({len(test_results)} tests):")
        for i, result in enumerate(test_results, 1):
            status = result.get('status', 'Unknown')
            duration = result.get('durationSeconds', 0)
            print(f"  {i}. Status: {status}, Duration: {duration}s")
            
            if result.get('recordingUrl'):
                print(f"     🎵 Recording: {result['recordingUrl']}")
            
            if result.get('transcriptionDataUrl'):
                print(f"     📝 Transcript: {result['transcriptionDataUrl']}")
    
    print("="*60)

async def wait_for_completion(api, test_run_id: str, max_wait_seconds: int = 600):
    """Wait for test completion and return results."""
    print(f"\n⏳ Waiting for test run {test_run_id} to complete...")
    print("💡 Tip: You can skip waiting and check results later in the dashboard")
    
    start_time = time.time()
    
    while True:
        if time.time() - start_time > max_wait_seconds:
            print(f"⏰ Timeout reached ({max_wait_seconds}s). Check dashboard for results.")
            return None
        
        status = await api.get_test_run_status(test_run_id)
        
        if status["status"] in ["COMPLETED", "FAILED"]:
            print(f"✅ Test run {status['status'].lower()}")
            break
        
        print(f"🔄 Status: {status['status']} - waiting...")
        await asyncio.sleep(10)
    
    return await api.get_test_results(test_run_id)

async def main():
    """Run an advanced outbound call test with full workflow."""
    print("🚀 Starting Advanced Outbound Call Test")
    print("=" * 50)
    
    # Configuration - update these values
    api_key = os.getenv("HAMMING_API_KEY", "your-api-key-here")
    agent_id = os.getenv("HAMMING_AGENT_ID", "your-agent-id")
    tag_ids = ["default", "api-test"]  # Specify which test cases to run
    
    if api_key == "your-api-key-here" or agent_id == "your-agent-id":
        print("⚠️  Please set HAMMING_API_KEY and HAMMING_AGENT_ID environment variables")
        print("   Or update the values in the script")
        return
    
    try:
        async with HammingOutboundCallAPI(api_key) as api:
            print("✅ API client initialized")
            
            # Create test run
            print("\n🔄 Creating test run...")
            test_run_response = await api.create_test_run(
                agent_id=agent_id,
                test_name="Advanced Outbound API Test",
                tag_ids=tag_ids
            )
            
            test_run_id = test_run_response["testRunId"]
            assigned_numbers = test_run_response["assignedNumbers"]
            
            print(f"✅ Test run created successfully!")
            print(f"📊 Test Run ID: {test_run_id}")
            
            # Display assigned numbers
            print_assigned_numbers(assigned_numbers)
            
            # IMPORTANT: Now you need to call these numbers to trigger your agent
            # Uncomment the following section to see agent dispatch examples:
            
            # NEXT STEPS - HOW TO DISPATCH YOUR AGENT:
            # 1. Use your telephony system to call the assigned numbers above
            # 2. Your calls will connect to Hamming's test infrastructure  
            # 3. Hamming will then call your agent's webhook/phone number
            # 4. Your agent should handle the incoming call and respond appropriately
            
            # AGENT DISPATCH EXAMPLES:
            # • Twilio: client.calls.create(to=assigned_number, from_=your_number, webhook_url)
            # • Direct SIP: Configure your PBX to call the assigned numbers
            # • Manual: Call the numbers from your test phone
            # • Automated: Use your existing call dispatch system
            
            # TWILIO INTEGRATION EXAMPLE:
            # pip install twilio
            # from twilio.rest import Client
            # client = Client('your_account_sid', 'your_auth_token')
            # for assignment in assigned_numbers:
            #     call = client.calls.create(
            #         to=assignment['phoneNumber'],
            #         from_='+1234567890',  # Your Twilio number
            #         url='https://yourapp.com/voice'  # Your agent's webhook
            #     )
            
            # Ask user if they want to wait for completion
            print("\n🤔 Options:")
            print("  1. Wait for test completion (recommended for small tests)")
            print("  2. Exit now and check results later in dashboard")
            
            choice = input("\nEnter choice (1 or 2): ").strip()
            
            if choice == "1":
                # Monitor the test execution
                print("\n⏳ Monitoring test execution...")
                results = await wait_for_completion(api, test_run_id)
                if results:
                    print(f"\n🎉 Test completed!")
                    print(f"📊 Final Status: {results.get('status', 'Unknown')}")
                    
                    # Get detailed results
                    print("\n📊 Retrieving detailed results...")
                    
                    # Display results summary
                    print("\n" + "=" * 50)
                    print("📋 TEST RESULTS SUMMARY")
                    print("=" * 50)
                    
                    summary = results.get('summary', {})
                    stats = summary.get('stats', {})
                    
                    print(f"Total Calls: {stats.get('total', 0)}")
                    print(f"Successful Calls: {stats.get('completed', 0) - stats.get('failed', 0)}")
                    print(f"Failed Calls: {stats.get('failed', 0)}")
                    print(f"Pending Calls: {stats.get('pending', 0)}")
                    
                    # Show individual call results
                    call_results = results.get('results', [])
                    if call_results:
                        print(f"\n📞 INDIVIDUAL CALL RESULTS:")
                        for i, call in enumerate(call_results, 1):
                            print(f"\n  Call {i}:")
                            print(f"    Status: {call.get('status', 'N/A')}")
                            print(f"    Duration: {call.get('durationSeconds', 0):.1f}s")
                            
                            if call.get('recordingUrl'):
                                print(f"    🎵 Recording: {call['recordingUrl']}")
                            
                            if call.get('transcriptionDataUrl'):
                                print(f"    📝 Transcript: {call['transcriptionDataUrl']}")
                            
                            if call.get('analysis'):
                                analysis = call['analysis']
                                print(f"    Analysis Score: {analysis.get('overall_score', 'N/A')}")
                                if analysis.get('key_points'):
                                    print(f"    Key Points: {', '.join(analysis['key_points'][:3])}")
                    
                    # Save results to file
                    results_file = f"outbound_test_results_{test_run_id}.json"
                    with open(results_file, 'w') as f:
                        json.dump(results, f, indent=2)
                    
                    print(f"\n💾 Full results saved to: {results_file}")
                    print("\n🎉 Test completed successfully!")
            else:
                # Just show dashboard link
                dashboard_url = f"{api.base_url}/test-runs/{test_run_id}"
                print(f"\n✅ Test run created successfully!")
                print(f"📊 Dashboard URL: {dashboard_url}")
                print(f"📞 Call the numbers shown above to execute the tests")
                print(f"💡 Tip: Monitor progress in the dashboard")
        
    except Exception as e:
        print(f"\n❌ Error: {str(e)}")
        return 1
    
    return 0

if __name__ == "__main__":
    exit_code = asyncio.run(main())
    exit(exit_code)